README.md
## Adafruit Servo Driver ##

[![CodeClimate](https://codeclimate.com/github/chrisdambrosio/adafruit-servo-driver/badges/gpa.svg)](https://codeclimate.com/github/chrisdambrosio/adafruit-servo-driver)

### Description ###

A Ruby implementation of Adafruit's Python library for the Adafruit PCA9685 16-Channel
PWM Servo Driver for use with the Raspberry Pi.

### Installation ###

```
gem install adafruit-servo-driver
```

### Usage/Example ###

```ruby
require 'adafruit-servo-driver'

pwm = PWM.new(0x40, true)
pwm.set_pwm_freq(50)

channel = 0

3.times do
  pwm.set_pwm(channel, 0, 212)
  sleep(0.5)
  pwm.set_pwm(channel, 0, 412)
  sleep(0.5)
end
```

### Relevant Links: ###

Product Page:
http://www.adafruit.com/products/815

Original Adafruit Python Adafruit_PWM_Servo_Driver Library:
https://github.com/adafruit/Adafruit-Raspberry-Pi-Python-Code

Tutorial: Adafruit 16 Channel Servo Driver with Raspberry Pi
http://learn.adafruit.com/adafruit-16-channel-servo-driver-with-raspberry-pi


CHANGELOG.md
## 0.0.2 (2015-01-16)

  - Update to work with rev 1 or 2 Raspberry Pi


spec/spec_helper.rb
require_relative '../lib/adafruit-servo-driver'

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    # be_bigger_than(2).and_smaller_than(4).description
    #   # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #   # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Limits the available syntax to the non-monkey patched syntax that is recommended.
  # For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/i2c_device_spec.rb
include AdafruitServoDriver

describe I2CDevice do
  context 'A Raspberry Pi rev 1' do
    before do
      allow(File).to receive('exist?').with('/dev/i2c-0').and_return true
    end

    it 'detects the device path' do
      expect(I2CDevice.detect).to eq '/dev/i2c-0'
    end
  end

  context 'A Raspberry Pi rev 2' do
    before do
      allow(File).to receive('exist?').with('/dev/i2c-0').and_return false
      allow(File).to receive('exist?').with('/dev/i2c-1').and_return true
    end

    it 'detects the device path' do
      expect(I2CDevice.detect).to eq '/dev/i2c-1'
    end
  end

  context 'an unknown device' do
    before do
      allow(File).to receive('exist?').with('/dev/i2c-0').and_return false
      allow(File).to receive('exist?').with('/dev/i2c-1').and_return false
    end

    it 'raises a DeviceNotFoundError' do
      expect{I2CDevice.detect}.to raise_error DeviceNotFoundError
    end
  end
end


lib/adafruit-servo-driver/pwm.rb
require 'i2c/i2c'
require 'logger'
include AdafruitServoDriver

class PWM
  # Registers
  MODE1         = 0x00
  MODE2         = 0x01
  SUBADR1       = 0x02
  SUBADR2       = 0x03
  SUBADR3       = 0x04
  PRESCALE      = 0xFE
  LED0_ON_L     = 0x06
  LED0_ON_H     = 0x07
  LED0_OFF_L    = 0x08
  LED0_OFF_H    = 0x09
  ALL_LED_ON_L  = 0xFA
  ALL_LED_ON_H  = 0xFB
  ALL_LED_OFF_L = 0xFC
  ALL_LED_OFF_H = 0xFD

  # Bits
  RESTART            = 0x80
  SLEEP              = 0x10
  ALLCALL            = 0x01
  INVRT              = 0x10
  OUTDRV             = 0x04

  attr_writer :logger

  def initialize(address=0x40, debug=false)
    logger.level = Logger::DEBUG if debug

    @address = address
    @i2c = I2C.create(I2CDevice.detect)

    logger.debug 'Reseting PCA9685 MODE1 (without SLEEP) and MODE2'

    setAllPWM(0, 0)
    @i2c.write(@address, MODE2, OUTDRV)
    @i2c.write(@address, MODE1, ALLCALL)
    sleep(0.005)           # wait for oscillator

    mode1 = @i2c.read(@address, 8, MODE1).unpack('C').first
    @i2c.write(@address, MODE1, ALLCALL)
    mode1 = mode1 & ~SLEEP # wake up (reset sleep)
    @i2c.write(@address, MODE1, mode1)
    sleep(0.005)           # wait for oscillator
  end

  def set_pwm_freq(freq)
    logger.debug "Setting PWM frequency to #{freq} Hz"
    logger.debug "Estimated pre-scale: #{prescale_val}"

    prescale = (prescale_val + 0.5).floor

    logger.debug "Final pre-scale: #{prescale}"

    oldmode = @i2c.read(@address, 8, MODE1).unpack('C').first
    newmode = (oldmode & 0x7F) | 0x10    # sleep
    @i2c.write(@address, MODE1, newmode) # go to sleep
    @i2c.write(@address, PRESCALE, prescale)
    @i2c.write(@address, MODE1, oldmode)
    sleep(0.005)
    @i2c.write(@address, MODE1, oldmode | RESTART)
  end
  alias_method :setPWMFreq, :set_pwm_freq

  def set_pwm(channel, on, off)
    @i2c.write(@address, LED0_ON_L+4*channel, on & 0xFF)
    @i2c.write(@address, LED0_ON_H+4*channel, on >> 8)
    @i2c.write(@address, LED0_OFF_L+4*channel, off & 0xFF)
    @i2c.write(@address, LED0_OFF_H+4*channel, off >> 8)
  end
  alias_method :setPWM, :set_pwm

  def set_all_pwm(on, off)
    @i2c.write(@address, ALL_LED_ON_L, on & 0xFF)
    @i2c.write(@address, ALL_LED_ON_H, on >> 8)
    @i2c.write(@address, ALL_LED_OFF_L, off & 0xFF)
    @i2c.write(@address, ALL_LED_OFF_H, off >> 8)
  end
  alias_method :setAllPWM, :set_all_pwm

  private

  def logger
    @logger ||= Logger.new(STDOUT)
  end

  def prescale_val
    @prescale_val ||= calculate_prescale_val
  end

  def calculate_prescale_val
    prescaleval = 25000000.0    # 25MHz
    prescaleval /= 4096.0       # 12-bit
    prescaleval /= Float(freq)
    prescaleval -= 1.0
  end
end


adafruit-pwm-servo-driver.rb
require 'i2c/i2c'

class PWM
  # Registers
  MODE1         = 0x00
  MODE2         = 0x01
  SUBADR1       = 0x02
  SUBADR2       = 0x03
  SUBADR3       = 0x04
  PRESCALE      = 0xFE
  LED0_ON_L     = 0x06
  LED0_ON_H     = 0x07
  LED0_OFF_L    = 0x08
  LED0_OFF_H    = 0x09
  ALL_LED_ON_L  = 0xFA
  ALL_LED_ON_H  = 0xFB
  ALL_LED_OFF_L = 0xFC
  ALL_LED_OFF_H = 0xFD

  # Bits
  RESTART            = 0x80
  SLEEP              = 0x10
  ALLCALL            = 0x01
  INVRT              = 0x10
  OUTDRV             = 0x04

  def initialize(address=0x40, debug=false)
    # TODO detect pi revision
    @address = address
    @i2c = I2C.create('/dev/i2c-1')
    @debug = debug
    puts 'Reseting PCA9685 MODE1 (without SLEEP) and MODE2' if @debug
    setAllPWM(0, 0)
    @i2c.write(@address, MODE2, OUTDRV)
    @i2c.write(@address, MODE1, ALLCALL)
    sleep(0.005)           # wait for oscillator

    mode1 = @i2c.read(@address, 8, MODE1).unpack('C').first
    @i2c.write(@address, MODE1, ALLCALL)
    mode1 = mode1 & ~SLEEP # wake up (reset sleep)
    @i2c.write(@address, MODE1, mode1)
    sleep(0.005)           # wait for oscillator
  end

  def setPWMFreq(freq)
    prescaleval = 25000000.0    # 25MHz
    prescaleval /= 4096.0       # 12-bit
    prescaleval /= Float(freq)
    prescaleval -= 1.0
    if @debug
      puts "Setting PWM frequency to #{freq} Hz"
      puts "Estimated pre-scale: #{prescaleval}"
    end
    prescale = (prescaleval + 0.5).floor
    if @debug
      puts "Final pre-scale: #{prescale}"
    end

    oldmode = @i2c.read(@address, 8, MODE1).unpack('C').first
    newmode = (oldmode & 0x7F) | 0x10    # sleep
    @i2c.write(@address, MODE1, newmode) # go to sleep
    @i2c.write(@address, PRESCALE, prescale)
    @i2c.write(@address, MODE1, oldmode)
    sleep(0.005)
    @i2c.write(@address, MODE1, oldmode | RESTART)
  end

  def setPWM(channel, on, off)
    @i2c.write(@address, LED0_ON_L+4*channel, on & 0xFF)
    @i2c.write(@address, LED0_ON_H+4*channel, on >> 8)
    @i2c.write(@address, LED0_OFF_L+4*channel, off & 0xFF)
    @i2c.write(@address, LED0_OFF_H+4*channel, off >> 8)
  end

  def setAllPWM(on, off)
    @i2c.write(@address, ALL_LED_ON_L, on & 0xFF)
    @i2c.write(@address, ALL_LED_ON_H, on >> 8)
    @i2c.write(@address, ALL_LED_OFF_L, off & 0xFF)
    @i2c.write(@address, ALL_LED_OFF_H, off >> 8)
  end
end
